File: examples/python/appwrite/services/storage.py
from ..service import Service
from ..exception import AppwriteException

class Storage(Service):

    def __init__(self, client):
        super(Storage, self).__init__(client)

    def list_files(self, search = None, limit = None, offset = None, order_type = None):
        """List Files"""

        params = {}
        path = '/storage/files'

        if search is not None: 
            params['search'] = search

        if limit is not None: 
            params['limit'] = limit

        if offset is not None: 
            params['offset'] = offset

        if order_type is not None: 
            params['orderType'] = order_type

        return self.client.call('get', path, {
            'content-type': 'application/json',
        }, params)

    def create_file(self, file, read = None, write = None):
        """Create File"""

        if file is None: 
            raise AppwriteException('Missing required parameter: "file"')

        params = {}
        path = '/storage/files'

        if file is not None: 
            params['file'] = file

        if read is not None: 
            params['read'] = read

        if write is not None: 
            params['write'] = write

        return self.client.call('post', path, {
            'content-type': 'multipart/form-data',
        }, params)

    def get_file(self, file_id):
        """Get File"""

        if file_id is None: 
            raise AppwriteException('Missing required parameter: "file_id"')

        params = {}
        path = '/storage/files/{fileId}'
        path = path.replace('{fileId}', file_id)                

        return self.client.call('get', path, {
            'content-type': 'application/json',
        }, params)

    def update_file(self, file_id, read, write):
        """Update File"""

        if file_id is None: 
            raise AppwriteException('Missing required parameter: "file_id"')

        if read is None: 
            raise AppwriteException('Missing required parameter: "read"')

        if write is None: 
            raise AppwriteException('Missing required parameter: "write"')

        params = {}
        path = '/storage/files/{fileId}'
        path = path.replace('{fileId}', file_id)                

        if read is not None: 
            params['read'] = read

        if write is not None: 
            params['write'] = write

        return self.client.call('put', path, {
            'content-type': 'application/json',
        }, params)

    def delete_file(self, file_id):
        """Delete File"""

        if file_id is None: 
            raise AppwriteException('Missing required parameter: "file_id"')

        params = {}
        path = '/storage/files/{fileId}'
        path = path.replace('{fileId}', file_id)                

        return self.client.call('delete', path, {
            'content-type': 'application/json',
        }, params)

    def get_file_download(self, file_id):
        """Get File for Download"""

        if file_id is None: 
            raise AppwriteException('Missing required parameter: "file_id"')

        params = {}
        path = '/storage/files/{fileId}/download'
        path = path.replace('{fileId}', file_id)                

        return self.client.call('get', path, {
            'content-type': 'application/json',
        }, params)

    def get_file_preview(self, file_id, width = None, height = None, gravity = None, quality = None, border_width = None, border_color = None, border_radius = None, opacity = None, rotation = None, background = None, output = None):
        """Get File Preview"""

        if file_id is None: 
            raise AppwriteException('Missing required parameter: "file_id"')

        params = {}
        path = '/storage/files/{fileId}/preview'
        path = path.replace('{fileId}', file_id)                

        if width is not None: 
            params['width'] = width

        if height is not None: 
            params['height'] = height

        if gravity is not None: 
            params['gravity'] = gravity

        if quality is not None: 
            params['quality'] = quality

        if border_width is not None: 
            params['borderWidth'] = border_width

        if border_color is not None: 
            params['borderColor'] = border_color

        if border_radius is not None: 
            params['borderRadius'] = border_radius

        if opacity is not None: 
            params['opacity'] = opacity

        if rotation is not None: 
            params['rotation'] = rotation

        if background is not None: 
            params['background'] = background

        if output is not None: 
            params['output'] = output

        return self.client.call('get', path, {
            'content-type': 'application/json',
        }, params)

    def get_file_view(self, file_id):
        """Get File for View"""

        if file_id is None: 
            raise AppwriteException('Missing required parameter: "file_id"')

        params = {}
        path = '/storage/files/{fileId}/view'
        path = path.replace('{fileId}', file_id)                

        return self.client.call('get', path, {
            'content-type': 'application/json',
        }, params)

File: examples/ruby/lib/appwrite/services/storage.rb
module Appwrite
    class Storage < Service

        def list_files(search: nil, limit: nil, offset: nil, order_type: nil)
            path = '/storage/files'

            params = {}

            if !search.nil?
                params[:search] = search
            end

            if !limit.nil?
                params[:limit] = limit
            end

            if !offset.nil?
                params[:offset] = offset
            end

            if !order_type.nil?
                params[:orderType] = order_type
            end

            return @client.call('get', path, {
                'content-type' => 'application/json',
            }, params);
        end

        def create_file(file:, read: nil, write: nil)
            if file.nil?
                raise Appwrite::Exception.new('Missing required parameter: "file"')
            end

            path = '/storage/files'

            params = {}

            if !file.nil?
                params[:file] = file
            end

            if !read.nil?
                params[:read] = read
            end

            if !write.nil?
                params[:write] = write
            end

            return @client.call('post', path, {
                'content-type' => 'multipart/form-data',
            }, params);
        end

        def get_file(file_id:)
            if file_id.nil?
                raise Appwrite::Exception.new('Missing required parameter: "fileId"')
            end

            path = '/storage/files/{fileId}'
                .gsub('{fileId}', file_id)

            params = {}

            return @client.call('get', path, {
                'content-type' => 'application/json',
            }, params);
        end

        def update_file(file_id:, read:, write:)
            if file_id.nil?
                raise Appwrite::Exception.new('Missing required parameter: "fileId"')
            end

            if read.nil?
                raise Appwrite::Exception.new('Missing required parameter: "read"')
            end

            if write.nil?
                raise Appwrite::Exception.new('Missing required parameter: "write"')
            end

            path = '/storage/files/{fileId}'
                .gsub('{fileId}', file_id)

            params = {}

            if !read.nil?
                params[:read] = read
            end

            if !write.nil?
                params[:write] = write
            end

            return @client.call('put', path, {
                'content-type' => 'application/json',
            }, params);
        end

        def delete_file(file_id:)
            if file_id.nil?
                raise Appwrite::Exception.new('Missing required parameter: "fileId"')
            end

            path = '/storage/files/{fileId}'
                .gsub('{fileId}', file_id)

            params = {}

            return @client.call('delete', path, {
                'content-type' => 'application/json',
            }, params);
        end

        def get_file_download(file_id:)
            if file_id.nil?
                raise Appwrite::Exception.new('Missing required parameter: "fileId"')
            end

            path = '/storage/files/{fileId}/download'
                .gsub('{fileId}', file_id)

            params = {}

            return @client.call('get', path, {
                'content-type' => 'application/json',
            }, params);
        end

        def get_file_preview(file_id:, width: nil, height: nil, gravity: nil, quality: nil, border_width: nil, border_color: nil, border_radius: nil, opacity: nil, rotation: nil, background: nil, output: nil)
            if file_id.nil?
                raise Appwrite::Exception.new('Missing required parameter: "fileId"')
            end

            path = '/storage/files/{fileId}/preview'
                .gsub('{fileId}', file_id)

            params = {}

            if !width.nil?
                params[:width] = width
            end

            if !height.nil?
                params[:height] = height
            end

            if !gravity.nil?
                params[:gravity] = gravity
            end

            if !quality.nil?
                params[:quality] = quality
            end

            if !border_width.nil?
                params[:borderWidth] = border_width
            end

            if !border_color.nil?
                params[:borderColor] = border_color
            end

            if !border_radius.nil?
                params[:borderRadius] = border_radius
            end

            if !opacity.nil?
                params[:opacity] = opacity
            end

            if !rotation.nil?
                params[:rotation] = rotation
            end

            if !background.nil?
                params[:background] = background
            end

            if !output.nil?
                params[:output] = output
            end

            return @client.call('get', path, {
                'content-type' => 'application/json',
            }, params);
        end

        def get_file_view(file_id:)
            if file_id.nil?
                raise Appwrite::Exception.new('Missing required parameter: "fileId"')
            end

            path = '/storage/files/{fileId}/view'
                .gsub('{fileId}', file_id)

            params = {}

            return @client.call('get', path, {
                'content-type' => 'application/json',
            }, params);
        end


        protected

        private
    end 
end